### PR TITLE
Refactor low level clients

### PR DESCRIPTION
**Pull request by ionut-arm**
This commit includes a few changes:
* remove explicit wire protocol version handling, following a similar
change in the interface
* add a provider ID field to the list_opcodes operation, with the
request being serviced by the core provider
* include an implicit provider ID within the basic client to allow it to
maintain a consistent operation while allowing it to expose an interface
that essentially covers the wire protocol contracts
* add checks that core operations are only performed on the core client
and cryptographic operations cannot be performed on the core client

Co-authored-by: Ionut Mihalcea <ionut.mihalcea@arm.com>
Co-authored-by: Hugues de Valon <hugues.devalon@arm.com>

Signed-off-by: Ionut Mihalcea <ionut.mihalcea@arm.com>

### CODE REVIEW

File: src/core/mod.rs [hunk 1]
```diff
@@ -468,10 +478,17 @@ impl CoreClient {
...
             NativeOperation::PsaVerifyHash(op),
-            provider.id(),
+            self.implicit_provider,
             &self.auth_data,
         )?;
 
         Ok(())
     }
+
+    fn can_provide_crypto(&self) -> Result<()> {
```
Comment (hug-dev) [inline src/core/mod.rs:488]: Nice one

File: src/core/operation_client.rs [hunk 2]
```diff
@@ -1,28 +1,35 @@
 // Copyright 2020 Contributors to the Parsec project.
 // SPDX-License-Identifier: Apache-2.0
+//! Operation-level client
 #![allow(dead_code)]
 
-use super::request_handler::RequestHandler;
+use super::request_client::RequestHandler;
```
Comment (hug-dev) [inline src/core/operation_client.rs:6]: and this one `RequestClient` as well 😬 ?

File: src/core/operation_client.rs [hunk 3]
```diff
@@ -107,34 +112,26 @@ impl OperationHandler {
 impl Default for OperationHandler {
     fn default() -> Self {
         OperationHandler {
-            converter: Box::from(ProtobufConverter {}),
-            wire_protocol_version_maj: 1,
-            wire_protocol_version_min: 0,
-            content_type: BodyType::Protobuf,
-            accept_type: BodyType::Protobuf,
+            content_type: Box::from(ProtobufConverter {}),
```
Comment (hug-dev) [inline src/core/operation_client.rs:115]: Maybe the field name `content_converter` is more appropriate?

File: src/core/operation_client.rs [hunk 4]
```diff
@@ -107,34 +112,26 @@ impl OperationHandler {
...
+    pub fn set_request_content_type(&mut self, content_type: Box<dyn Convert>) {
         self.op_handler.content_type = content_type;
-        self.op_handler.accept_type = content_type;
-        match content_type {
-            BodyType::Protobuf => self.op_handler.converter = Box::from(ProtobufConverter {}),
-        }
     }
 
-    /// Set the wire protocol version numbers to be used by the client.
+    /// Set the content type for responses handled by this client.
```
Comment (hug-dev) [inline src/core/operation_client.rs:131]: "accept" type?

**Comment by ionut-arm** [comment]
>     * should we keep the `core` folder? I think it makes sense in terms of abstraction: it is the core based on which any higher level client can be built on.

I think so, it makes sense to have the core segregated, especially if we don't want people to use this too much, we can document it accordingly

>     * you would normally expect the main structure to be contained in `core/mod.rs` but since our `BasicClient` has no "core" in it and to align it with the other ones, would it be nicer to have `basic_client.rs` as a separate file, like operation and request?

That makes sense!

>     * we have `ipc_client` in `core`, it might be confusing. Maybe it is worth renaming it just `ipc` or `ipc_connectors`?

Yeah, I noticed that too, will try and think of the proper name for it.

File: src/core/mod.rs [hunk 5]
```diff
@@ -2,13 +2,13 @@
...
-mod operation_handler;
-mod request_handler;
+pub mod operation_client;
+pub mod request_client;
 mod testing;
 
 use crate::auth::AuthenticationData;
 use crate::error::{ClientErrorKind, Error, Result};
-use operation_handler::OperationHandler;
+use operation_client::OperationHandler;
```
[the remark this answers is not in the document]
Comment (ionut-arm) [reply src/core/mod.rs:11]: Yep! It's a bit weird naming all of them clients, but alas... it makes more sense to keep this consistent

File: src/core/mod.rs [hunk 6]
```diff
@@ -190,18 +157,20 @@ impl Provider {
...
+pub struct BasicClient {
     op_handler: OperationHandler,
     auth_data: AuthenticationData,
+    implicit_provider: ProviderID,
 }
 
 /// Main client functionality.
-impl CoreClient {
+impl BasicClient {
     /// Create a new Parsec client given the authentication data of the app.
```
[the remark this answers is not in the document]
Comment (ionut-arm) [reply src/core/mod.rs:168]: Yep, good point, I tried to update the docs but I forgot about this one!

File: src/core/mod.rs [hunk 7]
```diff
@@ -190,18 +157,20 @@ impl Provider {
 /// in order to figure out if their desired use case and provider are
 /// available.
```
[the remark this answers is not in the document]
Comment (ionut-arm) [reply src/core/mod.rs:158]: Will try and think of a way to "mark" them somehow in the docs!

File: src/core/operation_client.rs (hunk 3 above)
Comment (ionut-arm) [reply src/core/operation_client.rs:115]: Yeah, I guess, especially since it's public

File: src/core/basic_client.rs [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+}
+
+/// Main client functionality.
+impl BasicClient {
+    /// Create a new Parsec client given the authentication data of the app and a provider ID.
+    ///
+    /// The `implicit_provider` will be used for all non-core operations. Thus, until a
+    /// different provider ID is set using [`set_implicit_provider`](#method.set_implicit_provider),
+    /// all cryptographic operations will be executed in the context of this provider (if it is
+    /// available within the service).
```
Comment (hug-dev) [inline src/core/basic_client.rs:166]: Nice, I did not know you could link with methods!

**Pull request merged**